[plot] Exception in stats widget when adding an image

With python3.4, PyQt5 on debian 8, I get an exception with the stats widget when adding an image.

Code to reproduce:
```
from silx import sx
w = sx.plot((1, 2, 3))
# Open the stats widgets
w.addImage(numpy.random.random(100).reshape(10, 10))
```

Exception raised:
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
~/src/silx/build/lib.linux-x86_64-3.4/silx/gui/plot/StatsWidget.py in _plotContentChanged(self, action, kind, legend)
    551             return
    552         if action == 'add':
--> 553             self._addItem(self.plot.getCurve(legend))
    554         elif action == 'remove':
    555             self._removeItem(legend, kind)

~/src/silx/build/lib.linux-x86_64-3.4/silx/gui/plot/StatsWidget.py in _addItem(self, item)
    386 
    387     def _addItem(self, item):
--> 388         assert isinstance(item, self.COMPATIBLE_ITEMS)
    389         if (item.getLegend(), self._getKind(item)) in self._lgdAndKindToItems:
    390             self._updateStats(item, self._getKind(item))

AssertionError: 
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
~/src/silx/build/lib.linux-x86_64-3.4/silx/gui/plot/StatsWidget.py in _plotContentChanged(self, action, kind, legend)
    551             return
    552         if action == 'add':
--> 553             self._addItem(self.plot.getCurve(legend))
    554         elif action == 'remove':
    555             self._removeItem(legend, kind)

~/src/silx/build/lib.linux-x86_64-3.4/silx/gui/plot/StatsWidget.py in _addItem(self, item)
    386 
    387     def _addItem(self, item):
--> 388         assert isinstance(item, self.COMPATIBLE_ITEMS)
    389         if (item.getLegend(), self._getKind(item)) in self._lgdAndKindToItems:
    390             self._updateStats(item, self._getKind(item))

AssertionError: 
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
~/src/silx/build/lib.linux-x86_64-3.4/silx/gui/plot/StatsWidget.py in _plotContentChanged(self, action, kind, legend)
    551             return
    552         if action == 'add':
--> 553             self._addItem(self.plot.getCurve(legend))
    554         elif action == 'remove':
    555             self._removeItem(legend, kind)

~/src/silx/build/lib.linux-x86_64-3.4/silx/gui/plot/StatsWidget.py in _addItem(self, item)
    386 
    387     def _addItem(self, item):
--> 388         assert isinstance(item, self.COMPATIBLE_ITEMS)
    389         if (item.getLegend(), self._getKind(item)) in self._lgdAndKindToItems:
    390             self._updateStats(item, self._getKind(item))

AssertionError: 
Out[6]: 'Unnamed Image 1.1'
```

--- a/silx/gui/plot/StatsWidget.py
+++ b/silx/gui/plot/StatsWidget.py
@@ -254,7 +254,7 @@ class StatsTable(TableWidget):
         self._columns = self._columns_index.keys()
         self.setColumnCount(len(self._columns))
 
-        self._updateItemObserve(init=True)
+        self._updateItemObserve()
         self._updateAllStats()
 
     def getStatsHandler(self):
@@ -286,21 +286,16 @@ class StatsTable(TableWidget):
         :rtype: :class:`.PlotWidget`
         """
         if self.plot:
-            self.plot.sigContentChanged.disconnect(self._plotContentChanged)
-            self.plot.sigPlotSignal.disconnect(self._zoomPlotChanged)
+            self._dealWithPlotConnection(create=False)
         self.plot = plot
         self.clear()
         if self.plot:
-            self.plot.sigPlotSignal.connect(self._zoomPlotChanged)
-            self._updateItemObserve(init=True)
+            self._dealWithPlotConnection(create=True)
+            self._updateItemObserve()
 
-    def _updateItemObserve(self, switchItemsDisplayedType=False, init=False):
+    def _updateItemObserve(self):
         if self.plot:
-            if init is False:
-                self._dealWithPlotConnection(create=False)
             self.clear()
-            if switchItemsDisplayedType:
-                self._displayOnlyActItem = not self._displayOnlyActItem
             if self._displayOnlyActItem is True:
                 activeCurve = self.plot.getActiveCurve(just_legend=False)
                 activeScatter = self.plot._getActiveItem(kind='scatter',
@@ -323,30 +318,41 @@ class StatsTable(TableWidget):
                                                  just_legend=False,
                                                  withhidden=True)
                 [self._addItem(histogram) for histogram in histograms]
-                self.plot.sigContentChanged.connect(self._plotContentChanged)
-            self._dealWithPlotConnection(create=True)
 
     def _dealWithPlotConnection(self, create=True):
+        """
+        Manage connection to plot signals
+
+        Note: connection on Item are managed by the _removeItem function
+        """
+        if self.plot is None:
+            return
         if self._displayOnlyActItem:
-            if self.callbackImage is None:
-                self.callbackImage = functools.partial(self._activeItemChanged, 'image')
-                self.callbackScatter = functools.partial(self._activeItemChanged,
-                                                         'scatter')
-                self.callbackCurve = functools.partial(self._activeItemChanged, 'curve')
             if create is True:
+                if self.callbackImage is None:
+                    self.callbackImage = functools.partial(self._activeItemChanged, 'image')
+                    self.callbackScatter = functools.partial(self._activeItemChanged, 'scatter')
+                    self.callbackCurve = functools.partial(self._activeItemChanged, 'curve')
                 self.plot.sigActiveImageChanged.connect(self.callbackImage)
                 self.plot.sigActiveScatterChanged.connect(self.callbackScatter)
                 self.plot.sigActiveCurveChanged.connect(self.callbackCurve)
             else:
-                self.plot.sigActiveImageChanged.disconnect(self.callbackImage)
-                self.plot.sigActiveScatterChanged.disconnect(self.callbackScatter)
-                self.plot.sigActiveCurveChanged.disconnect(self.callbackCurve)
+                if self.callbackImage is not None:
+                    self.plot.sigActiveImageChanged.disconnect(self.callbackImage)
+                    self.plot.sigActiveScatterChanged.disconnect(self.callbackScatter)
+                    self.plot.sigActiveCurveChanged.disconnect(self.callbackCurve)
+                self.callbackImage = None
+                self.callbackScatter = None
+                self.callbackCurve = None
         else:
             if create is True:
                 self.plot.sigContentChanged.connect(self._plotContentChanged)
             else:
                 self.plot.sigContentChanged.disconnect(self._plotContentChanged)
-                # Note: connection on Item arre managed by the _removeItem function
+        if create is True:
+            self.plot.sigPlotSignal.connect(self._zoomPlotChanged)
+        else:
+            self.plot.sigPlotSignal.disconnect(self._zoomPlotChanged)
 
     def clear(self):
         """
@@ -502,13 +508,17 @@ class StatsTable(TableWidget):
                 raise ValueError('kind not managed')
         qt.QTableWidget.currentChanged(self, current, previous)
 
-    def setDisplayOnlyActiveItem(self, b):
+    def setDisplayOnlyActiveItem(self, displayOnlyActItem):
         """
 
         :param bool b: True if we want to only show active item
         """
-        if self._displayOnlyActItem != b:
-            self._updateItemObserve(switchItemsDisplayedType=True)
+        if self._displayOnlyActItem == displayOnlyActItem:
+            return
+        self._displayOnlyActItem = displayOnlyActItem
+        self._dealWithPlotConnection(create=False)
+        self._updateItemObserve()
+        self._dealWithPlotConnection(create=True)
 
     def setStatsOnVisibleData(self, b):
         """
@@ -526,23 +536,7 @@ class StatsTable(TableWidget):
     def _activeItemChanged(self, kind):
         """Callback used when plotting only the active item"""
         assert kind in ('curve', 'image', 'scatter', 'histogram')
-
-        if kind == 'curve':
-            item = self.plot.getActiveCurve(just_legend=False)
-        elif kind == 'image':
-            item = self.plot.getActiveImage(just_legend=False)
-        elif kind == 'scatter':
-            item = self.plot._getActiveItem(kind='scatter', just_legend=False)
-        elif kind == 'histogram':
-            item = self.plot.getActiveHistogram(just_legend=False)
-        else:
-            raise ValueError('kind not managed')
-        if item is not None:
-            if (item.getLegend(), kind) not in self._lgdAndKindToItems:
-                self.clear()
-                self._addItem(item)
-            else:
-                self._updateCurrentStats()
+        self._updateItemObserve()
 
     def _plotContentChanged(self, action, kind, legend):
         """Callback used when plotting all the plot items"""

--- a/silx/gui/plot/StatsWidget.py
+++ b/silx/gui/plot/StatsWidget.py
@@ -387,7 +387,7 @@ class StatsTable(TableWidget):
     def _addItem(self, item):
         assert isinstance(item, self.COMPATIBLE_ITEMS)
         if (item.getLegend(), self._getKind(item)) in self._lgdAndKindToItems:
-            self._updateStats(item, self._getKind(item))
+            self._updateStats(item.getLegend(), self._getKind(item))
             return
 
         self.setRowCount(self.rowCount() + 1)
@@ -450,7 +450,7 @@ class StatsTable(TableWidget):
         for lgdAndKind in self._lgdAndKindToItems:
             self._updateStats(lgdAndKind[0], lgdAndKind[1])
 
-    def _updateStats(self, legend, kind, *args, **kwargs):
+    def _updateStats(self, legend, kind, event=None):
         if self._statsHandler is None:
             return
 
@@ -469,7 +469,6 @@ class StatsTable(TableWidget):
         if not item or (item.getLegend(), kind) not in self._lgdAndKindToItems:
             return
 
-        assert (item.getLegend(), kind) in self._lgdAndKindToItems
         assert isinstance(item, self.COMPATIBLE_ITEMS)
 
         statsValDict = self._statsHandler.calculate(item, self.plot,

--- a/silx/gui/plot/StatsWidget.py
+++ b/silx/gui/plot/StatsWidget.py
@@ -28,7 +28,7 @@ Module containing widgets displaying stats from items of a plot.
 
 __authors__ = ["H. Payno"]
 __license__ = "MIT"
-__date__ = "06/06/2018"
+__date__ = "12/06/2018"
 
 
 import functools
@@ -549,8 +549,21 @@ class StatsTable(TableWidget):
         """Callback used when plotting all the plot items"""
         if kind not in ('curve', 'image', 'scatter', 'histogram'):
             return
+        if kind == 'curve':
+            item = self.plot.getCurve(legend)
+        elif kind == 'image':
+            item = self.plot.getImage(legend)
+        elif kind == 'scatter':
+            item = self.plot.getScatter(legend)
+        elif kind == 'histogram':
+            item = self.plot.getHistogram(legend)
+        else:
+            raise ValueError('kind not managed')
+
         if action == 'add':
-            self._addItem(self.plot.getCurve(legend))
+            if item is None:
+                raise ValueError('Item from legend "%s" do not exists' % legend)
+            self._addItem(item)
         elif action == 'remove':
             self._removeItem(legend, kind)
 

